### PR TITLE
Feature/adapt existing optimizers to new history recording

### PR DESCRIPTION
**Pull request by dexter2206**
This adapts optimizers to the updated Optimizer interface. Basically, they now operate on callables (with gradient, where appropriate).

This is to be merged after https://github.com/zapatacomputing/z-quantum-core/pull/96

### CODE REVIEW

File: src/python/zquantum/optimizers/cma_es_optimizer.py [hunk 1]
```diff
@@ -32,13 +33,13 @@ def __init__(self, options):
                 "CMA-ES always keeps track of the history, regardless of the keep_value_history flag."
             )
 
-    def minimize(self, cost_function, initial_params):
+    def minimize(self, cost_function, initial_params: np.ndarray) -> OptimizeResult:
```
Comment (mstechly) [inline src/python/zquantum/optimizers/cma_es_optimizer.py:36]: Do we want to add typing here?

File: src/python/zquantum/optimizers/cma_es_optimizer_test.py [hunk 2]
```diff
@@ -14,7 +14,7 @@ def test_incorrect_initialization(self):
         self.assertRaises(TypeError, lambda: CMAESOptimizer())
 
     def test_cmaes_specific_fields(self):
-        cost_function = MockCostFunction()
+        cost_function = MockCostFunction().evaluate
```
Comment (mstechly) [inline src/python/zquantum/optimizers/cma_es_optimizer_test.py:17]: Aren't we getting rid of that? Or not yet?

File: src/python/zquantum/optimizers/cma_es_optimizer.py (hunk 1 above)
Comment (mstechly) [reply src/python/zquantum/optimizers/cma_es_optimizer.py:36]: I think the typing should be the same as in "grid_search" for example?

File: src/python/zquantum/optimizers/grid_search.py [hunk 3]
```diff
@@ -47,23 +48,20 @@ def minimize(
         min_value = None
         nfev = 0
 
+        if self.keep_value_history:
+            cost_function = recorder(cost_function)
```
Comment (mstechly) [inline src/python/zquantum/optimizers/grid_search.py:52]: I think that right now there's no way to set recorder settings without changing the code, right?
Shouldn't we have an input parameter/class field or sth. like this?

**Review by mstechly — CHANGES_REQUESTED** [review]

File: src/python/zquantum/optimizers/grid_search.py (hunk 3 above)
Comment (dexter2206) [reply src/python/zquantum/optimizers/grid_search.py:52]: You are right. I didn't want this PR to get too large to digest, so I focused only on reproducing current functionalities. Adding recorder settings (i.e. save condition) will be proposed in another PR, with a corresponding change in the core's Optimizer interface.

File: src/python/zquantum/optimizers/cma_es_optimizer_test.py (hunk 2 above)
Comment (dexter2206) [reply src/python/zquantum/optimizers/cma_es_optimizer_test.py:17]: That's the next (and the last step), but for now `CostFunction` interface is still present in core, along with the `MockCostFunction`.

Comment (mstechly) [reply src/python/zquantum/optimizers/cma_es_optimizer_test.py:17]: Ok 👍

File: src/python/zquantum/optimizers/grid_search.py (hunk 3 above)
Comment (mstechly) [reply src/python/zquantum/optimizers/grid_search.py:52]: Sounds good 👍

File: src/python/zquantum/optimizers/cma_es_optimizer.py (hunk 1 above)
Comment (dexter2206) [reply src/python/zquantum/optimizers/cma_es_optimizer.py]: I made it the same as in `grid_search`. The typing might need sone additional refinement, but it seems good to me for the time being.

**Review by mstechly — APPROVED** [review]

**Review by mstechly — APPROVED** [review]

**Review by mstechly — APPROVED** [review]

**Pull request merged**